src/toil/leader.py

```diff
@@ -583,29 +586,54 @@ def _processReadyJob(self, job_id: str, result_status: int):
                          "successful completion of its successor jobs",
                          readyJob)
             self.serviceManager.kill_services(self.toilState.servicesIssued[readyJob.jobStoreID], error=False)
         else:
-            #There are no remaining tasks to schedule within the job, but
-            #we schedule it anyway to allow it to be deleted. Remove the job
+            # There are no remaining tasks to schedule within the job.
+            #
+            # We want it to go away. Usually the worker is responsible for
+            # deleting completed jobs. But when the job has a
+            # child/followOn/service dependents that have to finish first, it
+            # gets kicked back here when they are done.
+            #
+            # We don't want to just issueJob() again because then we will
+            # forget that this job is actually done and not a retry. A
+            # successful-subtree checkpoint job can't be distinguished from a
+            # being-retried one.
+            #
+            # So we do the cleanup here, to avoid reserving any real resources.
+            # See: https://github.com/DataBiosphere/toil/issues/4158 and
+            # https://github.com/DataBiosphere/toil/issues/3188
+            #
+            # TODO: Does this want to happen in another thread? What if we
+            # don't manage to do all the jobstore file deletes and need to
+            # retry them? What if there are 1000 of them?
 
-            # TODO: resize down here so it doesn't schedule at full original size just to delete itself!
-
-            #TODO: An alternative would be simple delete it here and add it to the
-            #list of jobs to process, or (better) to create an asynchronous
-            #process that deletes jobs and then feeds them back into the set
-            #of jobs to be processed
             if readyJob.remainingTryCount > 0:
-                self.issueJob(readyJob)
-                logger.debug("Job: %s is empty, we are scheduling to clean it up", readyJob.jobStoreID)
+                # add attribute to let issueJob know that this is an empty job and should be deleted
+                logger.debug("Job: %s is empty, we are cleaning it up", readyJob)
+                
+                try:
+                    self.toilState.delete_job(readyJob.jobStoreID)
+                except Exception as e:
+                    logger.exception("Re-processing success for job we could not remove: %s", readyJob)
+                    # Kick it back to being handled as succeeded again. We
+                    # don't want to have a failure here cause a Toil-level
+                    # retry which causes more actual jobs to try to run.
+                    # TODO: If there's some kind of permanent failure deleting
+                    # from the job store, we will loop on this job forever!
+                    self.process_finished_job_description(readyJob, 0)
```

Comment:
Would we want Toil to loop forever in that case?

Reply:
Not really. That's why I put in the TODO. But I also couldn't really think of a good way to avoid it.